lib/rust/ensogl/component/grid-view/src/lib.rs

```diff
@@ -525,6 +538,29 @@ impl<E: Entry> GridView<E> {
         let widget = Widget::new(app, frp, model, display_object);
         Self { widget }
     }
+
+    fn move_selection_or_emit_movement_out_of_grid_prevented_event(
+        &self,
+        dir: frp::io::keyboard::ArrowKeyDirection,
+    ) {
+        use frp::io::keyboard::ArrowKeyDirection::*;
+        let frp = self.frp();
+        if let Some((row, col)) = frp.entry_selected.value() {
+            let (rows, cols) = frp.grid_size.value();
+            let new_selection_if_in_bounds = match dir {
+                Up if row > 0 => Some((row - 1, col)),
+                Down if row + 1 < rows => Some((row + 1, col)),
+                Left if col > 0 => Some((row, col - 1)),
+                Right if col + 1 < cols => Some((row, col + 1)),
```

Comment:
slightly repeated code, e.g. "row + 1". To be fixed. Also, you can simplify it by first computing new selection and then checking if it was in bounds.